modules/dem_uart/test/test_dem_uart.py

```diff
@@ -156,3 +84,139 @@ def test_dem_uart_activation(dut):
+
+        # out_char & out_valid must be set until both out_ready & out_valid
+        # are asserted at the same rising edge, and for at least one clock cycle 
+        yield RisingEdge(dut.clk)
+
+        while not dut.out_ready.value or not dut.out_valid:
+            yield RisingEdge(dut.clk)
+
+        # Negate out_valid after the module has read out_char
+        dut.out_valid <= 0
+
+@cocotb.coroutine
+def dem_to_dii_rx(dut, num_transfers=500, max_delay=5):
+
+    reader    = NocDiReader(dut, dut.clk);
+    ex_packet = DiPacket();
+
+    for i in range(num_transfers):
+
+        delay = random.randint(1, max(1, max_delay))
+
+        for _ in range(delay):
+            yield RisingEdge(dut.clk)
+
+        rxPacket = yield reader.receive_packet(set_ready=True)
+
+        data = _fifo.pop(0)
+
+        ex_packet.set_contents(dest=SENDER_DI_ADDRESS, src=MODULE_DI_ADDRESS,
+                               type=DiPacket.TYPE.EVENT, type_sub=0,
+                               payload=[int(data)])
+
+        if not rxPacket:
+            raise DebugEventFailedException("receive_packet() timed out")
+
+        if not rxPacket.equal_to(dut, ex_packet, mask=None):
+            raise DebugEventFailedException("Unexpected content of " +
+                                            rxPacket.__str__() + "\n Expected "
+                                            + ex_packet.__str__())
+
+@cocotb.test()
+def test_dii_to_dem(dut):
+    """
+    Verify that no chars are lost when the CPU is busy
+    """
+
+    yield _init_dut(dut)
+    yield _activate_module(dut)
+
+    write_thread = cocotb.fork(dii_to_dem_tx(dut, num_transfers=1000, random_data=True))
+    read_thread  = cocotb.fork(dii_to_dem_rx(dut, num_transfers=1000))
+
+    # read_thread implicitly waits for write thread to finish as well
+    yield read_thread.join()
+
+@cocotb.coroutine
+def dii_to_dem_tx(dut, num_transfers=500, max_delay=5, random_data=False):
+
+    writer = NocDiWriter(dut, dut.clk);
+    tx_packet = DiPacket();
+
+    for i in range(num_transfers):
+
+        delay = random.randint(1, max(1, max_delay))
+
+        for _ in range(delay):
+            yield RisingEdge(dut.clk)
+
+        data = (random.randint(0, 255) if random_data else 0x42)
+
+        _fifo.append(data)
+
+        tx_packet.set_contents(dest=MODULE_DI_ADDRESS, src=SENDER_DI_ADDRESS,
+                               type=DiPacket.TYPE.EVENT.value, type_sub=0,
+                               payload=[int(data)])
+
+        yield writer.send_packet(tx_packet)
+
+@cocotb.coroutine
+def dii_to_dem_rx(dut, num_transfers=500, max_delay=5):
+
+    for i in range(num_transfers):
+
+        delay = random.randint(0, max_delay)
+
+        for _ in range(delay):
+            yield RisingEdge(dut.clk)
+
+        dut.in_ready <= 0x1
+
+        # Again, make sure in_ready is asserted for at least one clock cycle
+        yield RisingEdge(dut.clk)
+
+        while not dut.in_valid.value or not dut.in_ready.value:
```

Comment:
Same here you are setting 'in_ready' yourself, so you don't have to check for it.